pyop2/mpi.py

```diff
@@ -118,63 +135,260 @@ def delcomm_outer(comm, keyval, icomm):
-        defaults to COMM_WORLD.
-
-    :returns: An mpi4py communicator."""
-    if comm_in is None:
-        comm_in = COMM_WORLD
-    if isinstance(comm_in, PETSc.Comm):
-        comm_in = comm_in.tompi4py()
-    elif not isinstance(comm_in, MPI.Comm):
-        raise ValueError("Don't know how to dup a %r" % type(comm_in))
-    if comm_in == MPI.COMM_NULL:
-        return comm_in
-    refcount = comm_in.Get_attr(refcount_keyval)
-    if refcount is not None:
-        # Passed an existing PyOP2 comm, return it
-        comm_out = comm_in
-        refcount[0] += 1
+    :arg comm: Communicator to query
+    """
+    global PYOP2_FINALIZED
+    if isinstance(comm, PETSc.Comm):
+        ispyop2comm = False
+    elif comm == MPI.COMM_NULL:
+        if PYOP2_FINALIZED is False:
+            raise ValueError("Communicator passed to is_pyop2_comm() is COMM_NULL")
+        else:
+            ispyop2comm = True
+    elif isinstance(comm, MPI.Comm):
+        ispyop2comm = bool(comm.Get_attr(refcount_keyval))
     else:
-        # Check if communicator has an embedded PyOP2 comm.
-        comm_out = comm_in.Get_attr(innercomm_keyval)
-        if comm_out is None:
-            # Haven't seen this comm before, duplicate it.
-            comm_out = comm_in.Dup()
-            comm_in.Set_attr(innercomm_keyval, comm_out)
-            comm_out.Set_attr(outercomm_keyval, comm_in)
-            # Refcount
-            comm_out.Set_attr(refcount_keyval, [1])
-            # Remember we need to destroy it.
-            dupped_comms.append(comm_out)
+        raise ValueError(f"Argument passed to is_pyop2_comm() is a {type(comm)}, which is not a recognised comm type")
+    return ispyop2comm
+
+
+def pyop2_comm_status():
+    """ Prints the reference counts for all comms PyOP2 has duplicated
+    """
+    status_string = 'PYOP2 Communicator reference counts:\n'
+    status_string += '| Communicator name                      | Count |\n'
+    status_string += '==================================================\n'
+    for comm in dupped_comms:
+        if comm == MPI.COMM_NULL:
+            null = 'COMM_NULL'
+            status_string += f'| {null:39}| {0:5d} |\n'
         else:
-            refcount = comm_out.Get_attr(refcount_keyval)
+            refcount = comm.Get_attr(refcount_keyval)[0]
             if refcount is None:
-                raise ValueError("Inner comm without a refcount")
-            refcount[0] += 1
-    return comm_out
+                refcount = -999
+            status_string += f'| {comm.name:39}| {refcount:5d} |\n'
+    return status_string
 
 
-# Comm used for compilation, stashed on the internal communicator
-compilationcomm_keyval = MPI.Comm.Create_keyval()
+class PyOP2Comm:
+    """ Use a PyOP2 internal communicator and
+    increment and decrement the internal comm.
+    :arg comm: Any communicator
+    """
+    def __init__(self, comm):
+        self.user_comm = comm
+        self.internal_comm = None
```

Comment:
I think it makes a bit more sense for a context manager whose purpose is to provide an internal communicator to hold onto a single internal communicator rather than creating them on the fly. I prefer `__del__`.